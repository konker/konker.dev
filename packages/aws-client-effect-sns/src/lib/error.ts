export const TAG = 'SnsError';

export type SnsError = {
  readonly _tag: typeof TAG;
  readonly _Params: unknown;
  readonly message: string;
  readonly cause: unknown;
};

export const toSnsError =
  <I>(params: I) =>
  (x: unknown): SnsError => {
    return {
      _tag: TAG,
      _Params: params,
      message: typeof x === 'object' && x && 'message' in x ? (x as any).message : String(x),
      cause: x,
    };
  };
