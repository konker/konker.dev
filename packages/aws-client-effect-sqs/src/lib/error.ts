export const TAG = 'SqsError';

export type SqsError = {
  readonly _tag: typeof TAG;
  readonly _Params: unknown;
  readonly message: string;
  readonly cause: unknown;
};

export const toSqsError =
  <I>(params: I) =>
  (x: unknown): SqsError => {
    return {
      _tag: TAG,
      _Params: params,
      message: typeof x === 'object' && x && 'message' in x ? (x as any).message : String(x),
      cause: x,
    };
  };
