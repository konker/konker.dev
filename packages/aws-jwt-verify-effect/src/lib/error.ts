export const TAG = 'JwtVerifyError';

export type JwtVerifyError = {
  readonly _tag: typeof TAG;
  readonly message: string;
  readonly cause: unknown;
};

export const toJwtVerifyError = (x: unknown): JwtVerifyError => {
  return {
    _tag: TAG,
    message: typeof x === 'object' && x && 'message' in x ? (x as any).message : String(x),
    cause: x,
  };
};
