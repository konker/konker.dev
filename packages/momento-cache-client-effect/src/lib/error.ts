import type { TinyError } from '@konker.dev/tiny-error-fp';
import { toTinyError } from '@konker.dev/tiny-error-fp';

export const ERROR_TAG = 'MomentoClientError' as const;
export type ERROR_TAG = typeof ERROR_TAG;

export type MomentoClientError = TinyError<ERROR_TAG>;
export const toMomentoClientError = toTinyError<ERROR_TAG>(ERROR_TAG);
