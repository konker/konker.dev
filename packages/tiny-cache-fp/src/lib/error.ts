import type { TinyError } from '@konker.dev/tiny-error-fp';
import { toTinyError } from '@konker.dev/tiny-error-fp';

export const ERROR_TAG = 'CacheError' as const;
export type ERROR_TAG = typeof ERROR_TAG;

export type CacheError = TinyError<ERROR_TAG>;
export const toCacheError = toTinyError<ERROR_TAG>(ERROR_TAG);
