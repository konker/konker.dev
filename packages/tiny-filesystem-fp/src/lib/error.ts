export const TAG = 'TinyFileSystemError';

export type TinyFileSystemError = {
  readonly _tag: typeof TAG;
  readonly message: string;
  readonly cause: unknown;
};

export function toTinyFileSystemError(x: unknown): TinyFileSystemError {
  return {
    _tag: TAG,
    message: typeof x === 'object' && x && 'message' in x ? (x as any).message : String(x),
    cause: x,
  };
}
