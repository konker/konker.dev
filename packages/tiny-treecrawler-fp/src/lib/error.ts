import { toTinyError } from '@konker.dev/tiny-error-fp';

export const ERROR_TAG = 'TinyTreeCrawlerError';
export type ERROR_TAG = typeof ERROR_TAG;

export const toTinyTreeCrawlerError = toTinyError<ERROR_TAG>(ERROR_TAG);
export type TinyTreeCrawlerError = ReturnType<typeof toTinyTreeCrawlerError>;
