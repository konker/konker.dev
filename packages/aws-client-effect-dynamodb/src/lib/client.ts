import * as dynamodb from '@aws-sdk/client-dynamodb';
import * as dynamodbDocClient from '@aws-sdk/lib-dynamodb';
import type { Client } from '@aws-sdk/types';
import { Context } from 'effect';
import * as Effect from 'effect/Effect';

import type { DynamoDbError } from './error.js';
import { toDynamoDbError } from './error.js';

//------------------------------------------------------
export type DynamoDBClientFactory = (config: dynamodb.DynamoDBClientConfig) => dynamodb.DynamoDBClient;
export const defaultDynamoDBClientFactory: DynamoDBClientFactory = (config: dynamodb.DynamoDBClientConfig) =>
  new dynamodb.DynamoDBClient(config);

//------------------------------------------------------
export type DynamoDBClientFactoryDeps = {
  readonly dynamoDBClientFactory: DynamoDBClientFactory;
};
export const DynamoDBClientFactoryDeps = Context.GenericTag<DynamoDBClientFactoryDeps>(
  'aws-client-effect-dynamodb/DynamoDBFactoryDeps'
);

export const defaultDynamoDBClientFactoryDeps = Effect.provideService(
  DynamoDBClientFactoryDeps,
  DynamoDBClientFactoryDeps.of({
    dynamoDBClientFactory: defaultDynamoDBClientFactory,
  })
);

//------------------------------------------------------
export type DynamoDBDocumentClientFactory = (
  dynamodbClient: dynamodb.DynamoDBClient
) => dynamodbDocClient.DynamoDBDocumentClient;

export const defaultDynamoDBDocumentClientFactory = (dynamodbClient: dynamodb.DynamoDBClient) =>
  dynamodbDocClient.DynamoDBDocumentClient.from(dynamodbClient);

//------------------------------------------------------
export type DynamoDBDocumentClientFactoryDeps = {
  readonly dynamoDBClientFactory: DynamoDBClientFactory;
  readonly dynamoDBDocumentClientFactory: DynamoDBDocumentClientFactory;
};
export const DynamoDBDocumentClientFactoryDeps = Context.GenericTag<DynamoDBDocumentClientFactoryDeps>(
  'aws-client-effect-dynamodb/DynamoDBDocumentClientFactoryDeps'
);

export const defaultDynamoDBDocumentClientFactoryDeps = Effect.provideService(
  DynamoDBDocumentClientFactoryDeps,
  DynamoDBDocumentClientFactoryDeps.of({
    dynamoDBClientFactory: defaultDynamoDBClientFactory,
    dynamoDBDocumentClientFactory: defaultDynamoDBDocumentClientFactory,
  })
);

//------------------------------------------------------
export type DynamoDBDocumentClientDeps = {
  readonly dynamoDBDocumentClient: Client<
    dynamodbDocClient.ServiceInputTypes,
    dynamodbDocClient.ServiceOutputTypes,
    unknown
  >;
  readonly dynamoDBClient: dynamodb.DynamoDBClient;
};
export const DynamoDBDocumentClientDeps = Context.GenericTag<DynamoDBDocumentClientDeps>(
  'aws-client-effect-dynamodb/DynamoDBDocumentClientDeps'
);

export const defaultDynamoDBDocumentClientDeps = (config: dynamodb.DynamoDBClientConfig) => {
  const defaultClientHandle = defaultDynamoDBClientFactory(config);
  const defaultDocClientHandle = defaultDynamoDBDocumentClientFactory(defaultClientHandle);

  return Effect.provideService(
    DynamoDBDocumentClientDeps,
    DynamoDBDocumentClientDeps.of({
      dynamoDBClient: defaultClientHandle,
      dynamoDBDocumentClient: defaultDocClientHandle,
    })
  );
};

//------------------------------------------------------
export const createDynamoDBDocumentClientDeps =
  (config: dynamodb.DynamoDBClientConfig) =>
  (factoryDeps: DynamoDBDocumentClientFactoryDeps): Effect.Effect<DynamoDBDocumentClientDeps, DynamoDbError> =>
    Effect.tryPromise({
      try: async () => {
        const clientHandle = factoryDeps.dynamoDBClientFactory(config);
        const docClientHandle = factoryDeps.dynamoDBDocumentClientFactory(clientHandle);
        return DynamoDBDocumentClientDeps.of({
          dynamoDBClient: clientHandle,
          dynamoDBDocumentClient: docClientHandle,
        });
      },
      catch: toDynamoDbError({}),
    });

//------------------------------------------------------
export const cleanupDynamoDBDocumentClientDeps = (deps: DynamoDBDocumentClientDeps) => () => {
  return Effect.tryPromise({
    // eslint-disable-next-line fp/no-nil
    try: async () => {
      // eslint-disable-next-line fp/no-unused-expression
      deps.dynamoDBDocumentClient.destroy();
      // eslint-disable-next-line fp/no-unused-expression
      deps.dynamoDBClient.destroy();
    },
    catch: toDynamoDbError({}),
  });
};
