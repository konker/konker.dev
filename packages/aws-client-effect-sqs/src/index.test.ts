import type {
  ChangeMessageVisibilityCommandInput,
  DeleteMessageBatchCommandInput,
  DeleteMessageCommandInput,
  ReceiveMessageCommandInput,
  SendMessageCommandInput,
} from '@aws-sdk/client-sqs';
import * as sqsClient from '@aws-sdk/client-sqs';
import { SQSClient } from '@aws-sdk/client-sqs';
import * as P from '@konker.dev/effect-ts-prelude';
import { mockClient } from 'aws-sdk-client-mock';
import { beforeAll, beforeEach, describe, expect, it } from 'vitest';

import * as unit from './index';
import { SQSClientDeps } from './index';
import { TAG } from './lib/error';

const sqsMock = mockClient(sqsClient.SQSClient);

describe('aws-client-effect-sqs', () => {
  // eslint-disable-next-line fp/no-let
  let deps: unit.SQSClientDeps;

  beforeAll(() => {
    // eslint-disable-next-line fp/no-mutation
    deps = unit.SQSClientDeps.of({
      sqsClient: new sqsClient.SQSClient({}),
    });
  });

  describe('Error tag', () => {
    it('should export as expected', () => {
      expect(unit.SQS_ERROR_TAG).toEqual('SqsError');
    });
  });

  // ------------------------------------------------------------------------
  describe('defaultSQSClientFactory', () => {
    it('should work as expected', async () => {
      expect(unit.defaultSQSClientFactory({})).toBeInstanceOf(SQSClient);
    });
  });

  // ------------------------------------------------------------------------
  describe('ChangeMessageVisibilityCommand', () => {
    beforeEach(() => {
      sqsMock.reset();
    });

    it('should work as expected', async () => {
      sqsMock.on(sqsClient.ChangeMessageVisibilityCommand).resolves({ $metadata: { httpStatusCode: 200 } });

      const params: ChangeMessageVisibilityCommandInput = {
        QueueUrl: 'https://test-queue-url',
        ReceiptHandle: 'test-receipt-handle',
        VisibilityTimeout: 123,
      };
      const expected = { $metadata: { httpStatusCode: 200 }, _Params: params };
      const command = P.pipe(
        unit.ChangeMessageVisibilityCommandEffect(params),
        P.Effect.provideService(SQSClientDeps, deps)
      );
      await expect(P.Effect.runPromise(command)).resolves.toStrictEqual(expected);
      expect(sqsMock.calls().length).toEqual(1);
    });

    it('should fail as expected', async () => {
      sqsMock.on(sqsClient.ChangeMessageVisibilityCommand).rejects({ $metadata: { httpStatusCode: 404 } });

      const params: ChangeMessageVisibilityCommandInput = {
        QueueUrl: 'https://test-queue-url',
        ReceiptHandle: 'test-receipt-handle',
        VisibilityTimeout: 123,
      };
      const command = P.pipe(
        unit.ChangeMessageVisibilityCommandEffect(params),
        P.Effect.provideService(SQSClientDeps, deps)
      );
      await expect(P.Effect.runPromise(command)).rejects.toThrowErrorMatchingSnapshot(JSON.stringify({ _tag: TAG }));
      expect(sqsMock.calls().length).toEqual(1);
    });
  });

  // ------------------------------------------------------------------------
  describe('DeleteMessageCommand', () => {
    beforeEach(() => {
      sqsMock.reset();
    });

    it('should work as expected', async () => {
      sqsMock.on(sqsClient.DeleteMessageCommand).resolves({ $metadata: { httpStatusCode: 200 } });

      const params: DeleteMessageCommandInput = {
        QueueUrl: 'https://test-queue-url',
        ReceiptHandle: 'test-receipt-handle',
      };
      const expected = { $metadata: { httpStatusCode: 200 }, _Params: params };
      const command = P.pipe(unit.DeleteMessageCommandEffect(params), P.Effect.provideService(SQSClientDeps, deps));
      await expect(P.Effect.runPromise(command)).resolves.toStrictEqual(expected);
      expect(sqsMock.calls().length).toEqual(1);
    });

    it('should fail as expected', async () => {
      sqsMock.on(sqsClient.DeleteMessageCommand).rejects({ $metadata: { httpStatusCode: 404 } });

      const params: DeleteMessageCommandInput = {
        QueueUrl: 'https://test-queue-url',
        ReceiptHandle: 'test-receipt-handle',
      };
      const command = P.pipe(unit.DeleteMessageCommandEffect(params), P.Effect.provideService(SQSClientDeps, deps));
      await expect(P.Effect.runPromise(command)).rejects.toThrowErrorMatchingSnapshot(JSON.stringify({ _tag: TAG }));
      expect(sqsMock.calls().length).toEqual(1);
    });
  });

  // ------------------------------------------------------------------------
  describe('DeleteMessageBatchCommand', () => {
    beforeEach(() => {
      sqsMock.reset();
    });

    it('should work as expected', async () => {
      sqsMock.on(sqsClient.DeleteMessageBatchCommand).resolves({ $metadata: { httpStatusCode: 200 } });

      const params: DeleteMessageBatchCommandInput = {
        QueueUrl: 'https://test-queue-url',
        Entries: [{ Id: 'test-id-1', ReceiptHandle: 'test-receipt-handle-1' }],
      };
      const expected = { $metadata: { httpStatusCode: 200 }, _Params: params };
      const command = P.pipe(
        unit.DeleteMessageBatchCommandEffect(params),
        P.Effect.provideService(SQSClientDeps, deps)
      );
      await expect(P.Effect.runPromise(command)).resolves.toStrictEqual(expected);
      expect(sqsMock.calls().length).toEqual(1);
    });

    it('should fail as expected', async () => {
      sqsMock.on(sqsClient.DeleteMessageBatchCommand).rejects({ $metadata: { httpStatusCode: 404 } });

      const params: DeleteMessageBatchCommandInput = {
        QueueUrl: 'https://test-queue-url',
        Entries: [{ Id: 'test-id-1', ReceiptHandle: 'test-receipt-handle-1' }],
      };
      const command = P.pipe(
        unit.DeleteMessageBatchCommandEffect(params),
        P.Effect.provideService(SQSClientDeps, deps)
      );
      await expect(P.Effect.runPromise(command)).rejects.toThrowErrorMatchingSnapshot(JSON.stringify({ _tag: TAG }));
      expect(sqsMock.calls().length).toEqual(1);
    });
  });

  // ------------------------------------------------------------------------
  describe('ReceiveMessageCommand', () => {
    beforeEach(() => {
      sqsMock.reset();
    });

    it('should work as expected', async () => {
      sqsMock
        .on(sqsClient.ReceiveMessageCommand)
        .resolves({ Messages: [{ MessageId: 'test-message-id' }], $metadata: { httpStatusCode: 200 } });

      const params: ReceiveMessageCommandInput = { QueueUrl: 'https://test-queue-url', VisibilityTimeout: 123 };
      const expected = {
        Messages: [{ MessageId: 'test-message-id' }],
        $metadata: { httpStatusCode: 200 },
        _Params: params,
      };
      const command = P.pipe(unit.ReceiveMessageCommandEffect(params), P.Effect.provideService(SQSClientDeps, deps));
      await expect(P.Effect.runPromise(command)).resolves.toStrictEqual(expected);
      expect(sqsMock.calls().length).toEqual(1);
    });

    it('should fail as expected', async () => {
      sqsMock.on(sqsClient.ReceiveMessageCommand).rejects({ $metadata: { httpStatusCode: 404 } });

      const params: ReceiveMessageCommandInput = { QueueUrl: 'https://test-queue-url', VisibilityTimeout: 123 };
      const command = P.pipe(unit.ReceiveMessageCommandEffect(params), P.Effect.provideService(SQSClientDeps, deps));
      await expect(P.Effect.runPromise(command)).rejects.toThrowErrorMatchingSnapshot(JSON.stringify({ _tag: TAG }));
      expect(sqsMock.calls().length).toEqual(1);
    });
  });

  // ------------------------------------------------------------------------
  describe('SendMessageCommand', () => {
    beforeEach(() => {
      sqsMock.reset();
    });

    it('should work as expected', async () => {
      sqsMock.on(sqsClient.SendMessageCommand).resolves({ MessageId: 'test-message-id' });

      const params: SendMessageCommandInput = { QueueUrl: 'https://test-queue-url', MessageBody: 'test-message' };
      const expected = { MessageId: 'test-message-id', _Params: params };
      const command = P.pipe(unit.SendMessageCommandEffect(params), P.Effect.provideService(SQSClientDeps, deps));
      await expect(P.Effect.runPromise(command)).resolves.toStrictEqual(expected);
      expect(sqsMock.calls().length).toEqual(1);
    });

    it('should fail as expected', async () => {
      sqsMock.on(sqsClient.SendMessageCommand).rejects({ $metadata: { httpStatusCode: 404 } });

      const params: SendMessageCommandInput = { QueueUrl: 'https://test-queue-url', MessageBody: 'test-message' };
      const command = P.pipe(unit.SendMessageCommandEffect(params), P.Effect.provideService(SQSClientDeps, deps));
      await expect(P.Effect.runPromise(command)).rejects.toThrowErrorMatchingSnapshot(JSON.stringify({ _tag: TAG }));
      expect(sqsMock.calls().length).toEqual(1);
    });
  });
});
