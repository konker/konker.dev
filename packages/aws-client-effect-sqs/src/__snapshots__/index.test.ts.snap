// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`aws-client-effect-sqs ChangeMessageVisibilityCommand should fail as expected: {"_tag":"SqsError"} 1`] = `"{"_tag":"SqsError","_Params":{"QueueUrl":"https://test-queue-url","ReceiptHandle":"test-receipt-handle","VisibilityTimeout":123},"message":"","cause":{"$metadata":{"httpStatusCode":404}}}"`;

exports[`aws-client-effect-sqs DeleteMessageBatchCommand should fail as expected: {"_tag":"SqsError"} 1`] = `"{"_tag":"SqsError","_Params":{"QueueUrl":"https://test-queue-url","Entries":[{"Id":"test-id-1","ReceiptHandle":"test-receipt-handle-1"}]},"message":"","cause":{"$metadata":{"httpStatusCode":404}}}"`;

exports[`aws-client-effect-sqs DeleteMessageCommand should fail as expected: {"_tag":"SqsError"} 1`] = `"{"_tag":"SqsError","_Params":{"QueueUrl":"https://test-queue-url","ReceiptHandle":"test-receipt-handle"},"message":"","cause":{"$metadata":{"httpStatusCode":404}}}"`;

exports[`aws-client-effect-sqs ReceiveMessageCommand should fail as expected: {"_tag":"SqsError"} 1`] = `"{"_tag":"SqsError","_Params":{"QueueUrl":"https://test-queue-url","VisibilityTimeout":123},"message":"","cause":{"$metadata":{"httpStatusCode":404}}}"`;

exports[`aws-client-effect-sqs SendMessageCommand should fail as expected: {"_tag":"SqsError"} 1`] = `"{"_tag":"SqsError","_Params":{"QueueUrl":"https://test-queue-url","MessageBody":"test-message"},"message":"","cause":{"$metadata":{"httpStatusCode":404}}}"`;
